src/Hazelcast.Net/CP/AtomicRef.cs

```diff
@@ -0,0 +1,90 @@
+        }
+
+        protected SerializationService SerializationService { get; }
+
+        /// <inheritdoc />
+        public async Task<bool> CompareAndSetAsync(T comparand, T value)
+        {
+            var requestMessage = AtomicRefCompareAndSetCodec.EncodeRequest(CPGroupId, Name, ToData(comparand), ToData(value));
+            var responseMessage = await Cluster.Messaging.SendAsync(requestMessage).CfAwait();
+            var response = AtomicRefCompareAndSetCodec.DecodeResponse(responseMessage).Response;
+            return response;
+        }
+
+        /// <inheritdoc />
+        public async Task<T> GetAsync()
+        {
+            var requestMessage = AtomicRefGetCodec.EncodeRequest(CPGroupId, Name);
+            var responseMessage = await Cluster.Messaging.SendAsync(requestMessage).CfAwait();
+            var response = AtomicRefGetCodec.DecodeResponse(responseMessage).Response;
+            return ToObject(response);
+        }
+
+        /// <inheritdoc />
+        public async Task SetAsync(T value)
+        {
+            var requestMessage = AtomicRefSetCodec.EncodeRequest(CPGroupId, Name, ToData(value), returnOldValue: false);
+            var responseMessage = await Cluster.Messaging.SendAsync(requestMessage).CfAwait();
+            var response = AtomicRefSetCodec.DecodeResponse(responseMessage).Response;
+        }
+
+        /// <inheritdoc />
+        public async Task<T> GetAndSetAsync(T value)
+        {
+            var requestMessage = AtomicRefSetCodec.EncodeRequest(CPGroupId, Name, ToData(value), returnOldValue: true);
+            var responseMessage = await Cluster.Messaging.SendAsync(requestMessage).CfAwait();
+            var response = AtomicRefSetCodec.DecodeResponse(responseMessage).Response;
+            return ToObject(response);
+        }
+
+        /// <inheritdoc />
+        public Task<bool> IsNullAsync() => ContainsAsync(null);
+
+        /// <inheritdoc />
+        public Task ClearAsync() => SetAsync(null);
+
+        /// <inheritdoc />
+        public async Task<bool> ContainsAsync(T value)
+        {
+            var requestMessage = AtomicRefContainsCodec.EncodeRequest(CPGroupId, Name, ToData(value));
+            var responseMessage = await Cluster.Messaging.SendAsync(requestMessage).CfAwait();
+            var response = AtomicRefContainsCodec.DecodeResponse(responseMessage).Response;
+            return response;
+        }
+
+        /// <inheritdoc />
+        public override async ValueTask DestroyAsync()
+        {
+            var requestMessage = CPGroupDestroyCPObjectCodec.EncodeRequest(CPGroupId, ServiceName, Name);
+            var responseMessage = await Cluster.Messaging.SendAsync(requestMessage).CfAwait();
+            var response = CPGroupDestroyCPObjectCodec.DecodeResponse(responseMessage);
+        }
+
+        protected IData ToData(T value) => SerializationService.ToData(value);
```

Comment:
Maybe we need to move this at `CPDistributedObjectBase` level?

Reply:
Thought about this but wanted to postpone `till more CP objects will be implemented as I'm not sure all of them relies on the same destroy API. What do you think?

Reply:
Ah, was talking about the `ToData` etc thing.

Reply:
This will require to move `SerializationService` instance to the `CPDistributedObjectBase`, but it looks like none of the CP objects except `AtomicReference` need it.